nodedb.go

```diff
@@ -671,6 +712,55 @@ func (ndb *nodeDB) decrVersionReaders(version int64) {
 	}
 }
 
+func (ndb *nodeDB) flushVersion(version int64) error {
+	// Create new recentDB and snapshotDB batch objects. We don't use the current
+	// batch objects of the nodeDB as we don't want to write state prematurely or
+	// have conflicting state.
+	//
+	// NOTE: We ignore any write that happen to recentBatch.
+	rb := ndb.recentDB.NewBatch()
+	defer rb.Close()
+
+	sb := ndb.snapshotDB.NewBatch()
+	defer sb.Close()
+
+	rootHash, err := ndb.getRoot(version)
+	if err != nil {
+		return err
+	}
+
+	switch {
+	case rootHash == nil:
+		return fmt.Errorf("version %v does not exist in recentDB", version)
+
+	case len(rootHash) == 0:
+		ndb.saveRootBatch([]byte{}, version, rb, sb)
+
+	default:
+		// save branch, the root, and the necessary orphans
+		node := ndb.GetNode(rootHash)
+		ndb.saveBranchBatch(node, true, rb, sb)
+		ndb.saveRootBatch(node.hash, version, rb, sb)
+	}
+
+	traverseOrphansVersionFromDB(ndb.recentDB, version, func(k, v []byte) {
+		var fromVersion, toVersion int64
+		orphanKeyFormat.Scan(k, &toVersion, &fromVersion)
+
+		ndb.saveOrphan(v, fromVersion, toVersion, true)
+	})
+
+	if err := sb.WriteSync(); err != nil {
+		return fmt.Errorf("failed to write (sync) the snapshot batch: %w", err)
+	}
+
+	if err := rb.WriteSync(); err != nil {
```

Comment:
what if this write fails, but `sb.WriteSync()` succeeds? will the IAVL end up in inconsistent state?

Reply:
Technically, we don't even need the `rb.WriteSync` call, because everything we should be writing to is the snapshotDB batch object. I'll remove this line for clarity.